libraries/chain/block_log.cpp

```diff
@@ -469,20 +478,243 @@ namespace eosio { namespace chain {
+         block_log_impl(const fc::path& data_dir, fc::path archive_dir, uint64_t stride, uint16_t max_retained_files);
+
+         static void ensure_file_exists(fc::cfile& f) {
+            if (fc::exists(f.get_file_path()))
+               return;
+            f.open(fc::cfile::create_or_update_rw_mode);
+            f.close();
+         }
+
+         uint64_t get_block_pos(uint32_t block_num);
+
+         void reset(uint32_t first_block_num, std::variant<genesis_state, chain_id_type>&& chain_context);
+
+         void flush();
+
+         uint64_t append(const signed_block_ptr& b, packed_transaction::cf_compression_type segment_compression);
+
+         uint64_t write_log_entry(const signed_block& b, packed_transaction::cf_compression_type segment_compression);
+
+         void split_log();
+         bool recover_from_incomplete_block_head(block_log_data& log_data, block_log_index& index);
+
+         block_id_type                 read_block_id_by_num(uint32_t block_num);
+         std::unique_ptr<signed_block> read_block_by_num(uint32_t block_num);
+         void                          read_head();
+      };
+   } // namespace detail
+
+   block_log::block_log(const fc::path& data_dir, fc::path archive_dir, uint64_t stride,
+                        uint16_t max_retained_files)
+       : my(new detail::block_log_impl(data_dir, archive_dir, stride, max_retained_files)) {}
+
+   block_log::~block_log() {}
+
+   bool detail::block_log_impl::recover_from_incomplete_block_head(block_log_data& log_data, block_log_index& index) {
+      if (preamble.version >= pruned_transaction_version) {
+         // check if the last block position
+         if (log_data.size() > index.back() + sizeof(uint32_t)) {
+            const uint32_t entry_size              = read_buffer<uint32_t>(log_data.data() + index.back());
+            const uint64_t trimmed_block_file_size = index.back() + entry_size;
+            const uint32_t expected_block_num      = log_data.first_block_num() + index.num_blocks() - 1;
+            if (log_data.size() > trimmed_block_file_size) {
+               try {
+                  log_data.light_validate_block_entry_at(index.back(), expected_block_num);
+                  ilog("The last block from blocks.log is incomplete, trim it.");
+                  boost::filesystem::resize_file(block_file.get_file_path(), trimmed_block_file_size);
+                  return true;
+               }
+               catch(...) {
+                  return false;
+               }
+            }
+         }
+      }
+      return false;
    }
 
-   block_log::~block_log(){}
+   detail::block_log_impl::block_log_impl(const fc::path& data_dir, fc::path archive_dir, uint64_t stride,
+                                          uint16_t max_retained_files) {
 
-   detail::block_log_impl::block_log_impl(const fc::path& data_dir) {
       if (!fc::is_directory(data_dir))
          fc::create_directories(data_dir);
+      else
+         catalog.open(data_dir);
```

Comment:
Possibly unpopular thought: discontinue v1 block log format support in nodeos.  Instead provide a utility to convert v1 format to v2 format which can be run manually prior to upgrading to this version of nodeos.

Otherwise need continue to maintain backwards support (forever)?

Reply:
test case added to split log from v1 to the latest version; i.e. if the system starts with v1 block log, it will  continue using the v1 log until it reaches the block number to split. After splitting, the newer block log would use the latest block log version regardless of the original block log version.